<a href="https://colab.research.google.com/github/minjeon99/ESAA-10th-OB/blob/Week5/w5_fri_study.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **08 텍스트 분석**

## **7. 문서 군집화 소개와 실습(Opinion Review 데이터 세트)**

### **문서 군집화 개념**
- **문서 군집화(Document Clustering)**: 비슷한 텍스트 구성의 문서를 군집화하기
  - 텍스트 분류 기반의 문서 분류와 유사하나, 문서 군집화는 학습 데이터 세트가 필요 없는 비지도학습 기반으로 동작

### **Opinion Review 데이터 세트를 이용한 문서 군집화 수행하기**
- **Opinion Review 데이터**: 51개의 텍스트 파일
  - Tripadvisor(호텔), Edmunds.com(자동차), Amazon.com(전자제품) 사이트에서 가져온 리뷰 문서
  - 각 문서는 약 100개 정도의 문장을 가짐
- 여러 개의 파일을 한 개의 DataFrame으로 로딩해 데이터 처리하기

In [8]:
import pandas as pd
import glob, os
import warnings
warnings.filterwarnings('ignore')
pd.set_option('display.max_colwidth', 700)

# 압축 파일을 풀어놓은 디렉터리
path = r'/content/Opinosis'
# path로 지정한 디렉터리 밑에 있는 모든 .data 파일들의 파일명을 리스트로 취합
all_files = glob.glob(os.path.join(path, "*.data"))
filename_list = []
opinion_text = []

# 개별 파일들의 파일명을 filename_list로 취합
# 개별 파일들의 파일 내용은 DataFrame으로 로딩 후 String으로 변환하여 opinion_text로 취합
for file_ in all_files:
  # 개별 파일을 읽어서 DataFrame으로 생성
  df = pd.read_table(file_, index_col=None, header=0, encoding='latin1')
  # 절대경로로 주어진 파일명을 가공
  # 맨 마지막 .data 확장자 제거
  filename_ = file_.split('/')[-1]
  filename = filename_.split('.')[0]

  # 파일명 리스트와 파일 내용 리스트에 파일명과 파일 내용을 추가
  filename_list.append(filename)
  opinion_text.append(df.to_string())

# 파일명 리스트와 파일 내용 리스트를 DataFrame으로 생성
document_df = pd.DataFrame({'filename':filename_list, 'opinion_text':opinion_text})
document_df.head()

filename  \
0  directions_garmin_nuvi_255W_gps   
1     display_garmin_nuvi_255W_gps   
2   satellite_garmin_nuvi_255W_gps   
3            screen_netbook_1005ha   
4             screen_ipod_nano_8gb   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  opinion_text  
0                                                                                                                                                                                                                                                                You also get upscale features like spoken directions including street names and programmable POIs .\n0                                                                                                                                                                                                                                                                                                I used to hesitate to go out of my directions but no...  
1                                                                                                                                                                                                                                                       3 quot  widescreen display was a bonus .\n0                                                                                                                                           This made for smoother graphics on the 255w of the vehicle moving along displayed roads, where the 750's display was more of a  jerky  movement .\n1                                                                                                                         ...  
2                                                                                                                                                                                                                                                                                          It's fast to acquire satellites .\n0    If you've ever had a  Brand X  GPS take you on some strange route that adds 20 minutes to your trip, has you turn the wrong way down a one way road, tell you to turn AFTER you've passed the street, frequently loses the satellite signal, or has old maps missing streets, you know how important this stuff is .\n1                                                                 ...  
3                                                                                                                                                                                                                                                                          Keep in mind that once you get in a room full of light or step outdoors screen reflections could become annoying .\n0                                                                                                                                                                                                                                                                                                       I've used mine outsi...  
4                                                                                                                                                                                                                                                                               

- TF-IDF 형태로 피처 벡터화하기
  - Lemmatization 같은 어근 변환을 직접 지원하진 않지만, tokenizer 인자에 커스텀 어근 변환 함수를 적용해 수행 가능

In [9]:
from nltk.stem import WordNetLemmatizer
import nltk
import string

remove_punct_dict = dict((ord(punct), None) for punct in string.punctuation)
lemmar = WordNetLemmatizer()

# 입력으로 들어온 token단어들에 대해서 lemmatization 어근 변환.
def LemTokens(tokens):
    return [lemmar.lemmatize(token) for token in tokens]

# TfidfVectorizer 객체 생성 시 tokenizer인자로 해당 함수를 설정하여 lemmatization 적용
# 입력으로 문장을 받아서 stop words 제거-> 소문자 변환 -> 단어 토큰화 -> lemmatization 어근 변환.
def LemNormalize(text):
    return LemTokens(nltk.word_tokenize(text.lower().translate(remove_punct_dict)))

In [15]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf_vect = TfidfVectorizer(tokenizer=LemNormalize, stop_words='english',
                             ngram_range=(1,2), min_df=0.05, max_df=0.85)

# opinion_text 칼럼값으로 feature vectorization 수행
feature_vect = tfidf_vect.fit_transform(document_df['opinion_text'])

- 문서별 텍스트가 TF-IDF 변환된 피처 벡터화 행렬 데이터에 대해 군집화 수행 -> 어떤 문서끼리 군집되는지 확인
  - K-Means 기법
- 문서의 유형: 크게 3가지(전자제품, 자동차, 호텔)
  - 세부 요소로 더 나뉨(전자제품 ->네비게이션, 아이팟, ...)
- 5개의 Centroid 기반으로 어떻게 군집화되는지 확인

In [16]:
from sklearn.cluster import KMeans

# 5개 집합으로 군집화
km_cluster = KMeans(n_clusters=5, max_iter=10000, random_state=0)
km_cluster.fit(feature_vect)
cluster_label = km_cluster.labels_
cluster_centers = km_cluster.cluster_centers_

- 각 데이터별로 할당된 군집의 레이블을 'cluster_label' 칼럼에 저장

In [17]:
document_df['cluster_label'] = cluster_label
document_df.head()

filename  \
0  directions_garmin_nuvi_255W_gps   
1     display_garmin_nuvi_255W_gps   
2   satellite_garmin_nuvi_255W_gps   
3            screen_netbook_1005ha   
4             screen_ipod_nano_8gb   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  opinion_text  \
0                                                                                                                                                                                                                                                                You also get upscale features like spoken directions including street names and programmable POIs .\n0                                                                                                                                                                                                                                                                                                I used to hesitate to go out of my directions but no...   
1                                                                                                                                                                                                                                                       3 quot  widescreen display was a bonus .\n0                                                                                                                                           This made for smoother graphics on the 255w of the vehicle moving along displayed roads, where the 750's display was more of a  jerky  movement .\n1                                                                                                                         ...   
2                                                                                                                                                                                                                                                                                          It's fast to acquire satellites .\n0    If you've ever had a  Brand X  GPS take you on some strange route that adds 20 minutes to your trip, has you turn the wrong way down a one way road, tell you to turn AFTER you've passed the street, frequently loses the satellite signal, or has old maps missing streets, you know how important this stuff is .\n1                                                                 ...   
3                                                                                                                                                                                                                                                                          Keep in mind that once you get in a room full of light or step outdoors screen reflections could become annoying .\n0                                                                                                                                                                                                                                                                                                       I've used mine outsi...   
4                                                                                                                                                                                                                                                                          

**Cluster 0**

In [18]:
document_df[document_df['cluster_label']==0].sort_values(by='filename')

filename  \
38       comfort_honda_accord_2008   
31       comfort_toyota_camry_2007   
26   gas_mileage_toyota_camry_2007   
48      interior_honda_accord_2008   
27      interior_toyota_camry_2007   
32       mileage_honda_accord_2008   
47   performance_honda_accord_2008   
8        quality_toyota_camry_2007   
40         seats_honda_accord_2008   
10  transmission_toyota_camry_2007   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
38                                                                                                                                                                                                                                                                                                                                                                                                Drivers seat not comfortable, the car itself compared to other models of similar class .\n0                                                                                                                                                                                                                               ...   
31                                                                                                                                                  Ride seems comfortable and gas mileage fairly good averaging 26 city and 30 open road .\n0                                                                                                        Seats are fine, in fact of all the smaller sedans this is the most comfortable I found for the price as I am 6', 2  and 250# .\n1                                                                                                                                                                                     Great gas mileage and comfortable on long trips ...   
26                                                                                                                                                                                                                                                                                                                                                                              Ride seems comfortable and gas mileage fairly good averaging 26 city and 30 open road .\n0                                                                                                                                                                                                                                                  ...   
48                                                                                                                                                                                                                                                                                                                                                                                                                                                                             I love the new body style and the interior is a simple pleasure except for the center dash .\n0                                                                                                                                              ...   
27                                                                      

- 자동차에 대한 리뷰
  - 토요타(Toyota), 혼다(Honda) 등

**Cluster 1**

In [19]:
document_df[document_df['cluster_label']==1].sort_values(by='filename')

filename  \
16   battery-life_amazon_kindle   
30   battery-life_ipod_nano_8gb   
17  battery-life_netbook_1005ha   
50      keyboard_netbook_1005ha   
45   performance_netbook_1005ha   
4          screen_ipod_nano_8gb   
3         screen_netbook_1005ha   
5      size_asus_netbook_1005ha   
36          sound_ipod_nano_8gb   
44          video_ipod_nano_8gb   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
16                                                                                                                                                                                                               After I plugged it in to my USB hub on my computer to charge the battery the charging cord design is very clever !\n0                                                                                                                                     After you have paged tru a 500, page book one, page, at, a, time to get from Chapter 2 to Chapter 15, see how excited you are about a low battery and all the time it took to get there !\n1                                                     ...   
30                                                                                                                                                                                                                                                                                                 short battery life  I moved up from an 8gb .\n0                                                                                                                                                                                                                                                                                            I love this ipod except for the battery life .\n1                             ...   
17                                                                                                                                                                                                                                                                                                                  6GHz 533FSB cpu, glossy display, 3, Cell 23Wh Li, ion Battery  , and a 1 .\n0                                                                                                                                                                                                                                                                                              Not to mention that as of now...   
50                                                                                                                                                                                  ,  I think the new keyboard rivals the great hp mini keyboards .\n0                                                                                                                                 Since the battery life difference is minimum, the only reason to upgrade would be to get the better keyboard .\n1                                                                                                                                                                                   The keyboard is now as good as t...   
45                                                                                                    

- 킨들, 아이팟, 넷북 등 포터블 전자기기

**Cluster 2**

In [20]:
document_df[document_df['cluster_label']==2].sort_values(by='filename')

filename  \
20    accuracy_garmin_nuvi_255W_gps   
19            buttons_amazon_kindle   
0   directions_garmin_nuvi_255W_gps   
1      display_garmin_nuvi_255W_gps   
14    eyesight-issues_amazon_kindle   
28                features_windows7   
15              fonts_amazon_kindle   
33         navigation_amazon_kindle   
2    satellite_garmin_nuvi_255W_gps   
7       screen_garmin_nuvi_255W_gps   
24       speed_garmin_nuvi_255W_gps   
35                   speed_windows7   
21     updates_garmin_nuvi_255W_gps   
6        voice_garmin_nuvi_255W_gps   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
20                                                                                                                                                                                                                                                                 , and is very, very accurate .\n0                                                                                                                                                                           but for the most part, we find that the Garmin software provides accurate directions, whereever we intend to go .\n1                                                                                                              This functi...   
19                                                                                                                                             I thought it would be fitting to christen my Kindle with the Stephen King novella UR, so went to the Amazon site on my computer and clicked on the button to buy it .\n0                                                                                                                                                                                                            As soon as I'd clicked the button to confirm my order it appeared on my Kindle almost immediately !\n1                                                                                   ...   
0                                                                                                                                                                                                                                                                 You also get upscale features like spoken directions including street names and programmable POIs .\n0                                                                                                                                                                                                                                                                                                I used to hesitate to go out of my directions but no...   
1                                                                                                                                                                                                                                                        3 quot  widescreen display was a bonus .\n0                                                                                                                                           This made for smoother graphics on the 255w of the vehicle moving along displayed roads, where the 750's display was more of a  jerky  movement .\n1                                  

- 주로 차량용내비게이션(garmin_nuvi)
  - 킨들등이 군집에 포함되어 있긴 함

**Cluster 3**

In [21]:
document_df[document_df['cluster_label']==3].sort_values(by='filename')

filename  \
42         food_holiday_inn_london   
34          food_swissotel_chicago   
13  location_bestwestern_hotel_sfo   
11     location_holiday_inn_london   
41   parking_bestwestern_hotel_sfo   
29             price_amazon_kindle   
49        price_holiday_inn_london   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
42                                                                                                                                                                                                                                                                      The room was packed to capacity with queues at the food buffets .\n0                                                                                                                                                   The over zealous   staff cleared our unfinished drinks while we were collecting cooked food and movement around the room with plates was difficult in the crowded circumstances .\n1                                         ...   
34                                                                                                                                                                                                                                                                                                                                                                   The food for our event was delicious .\n0                                                                                                                                                                                                                                                                                                              ...   
13                                                                                                                                                                                                                                                                                                       Good Value good location ,  ideal choice .\n0                                                                                                                                                                                                                                                                                          Great Location ,  Nice   Rooms ,  Helpless Concierge\n1                     ...   
11                                                                                                                                                                                                                                                                                                                  Great location for tube and we crammed in a fair amount of sightseeing in a short time .\n0                                                                                                                                                                                                                                                                All in all, a normal chain hotel on a nice lo...   
41                                                                                                                                                                                        

- 주로 호텔에 관한 리뷰

**Cluster 4**

In [22]:
document_df[document_df['cluster_label']==4].sort_values(by='filename')

filename  \
39   bathroom_bestwestern_hotel_sfo   
43       free_bestwestern_hotel_sfo   
23          room_holiday_inn_london   
12      rooms_bestwestern_hotel_sfo   
25          rooms_swissotel_chicago   
18    service_bestwestern_hotel_sfo   
22       service_holiday_inn_london   
46  service_swissotel_hotel_chicago   
37      staff_bestwestern_hotel_sfo   
9           staff_swissotel_chicago   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
39                                                                                                                                                                                                                                                The room was not overly big, but clean and very comfortable beds, a great shower and very clean bathrooms .\n0                                                                                                                                                                                                                          The second room was smaller, with a very inconvenient bathroom layout, but at least it was quieter and we were able to sleep .\n1 ...   
43                                                                                                                                                                The wine reception is a great idea as it is nice to meet other travellers and great having access to the free Internet access in our room .\n0                                                                                      They also have a computer available with free internet which is a nice bonus but I didn't find that out till the day before we left but was still able to get on there to check our flight to Vegas the next day .\n1                                                                                                 ...   
23                                                                                                                                                                                                                                                                                                                                                                                                                          We arrived at 23,30 hours and they could not recommend a restaurant so we decided to go to Tesco, with very limited choices but when you are hingry you do not careNext day they rang the bell at 8,00 hours to clean the room, not being very nice being waken up so earlyEvery day they gave u...   
12                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       Great Location ,  Nice   Rooms ,  H...   
25                                                            

- 주로 호텔에 관한 리뷰

전반적으로 군집 개수가 약간 많게 설정돼있어서 세분화되어 군집화된 경향
- 5개 -> 3개로 다시 군집화

In [23]:
from sklearn.cluster import KMeans

# 3개의 집합으로 군집화
km_cluster = KMeans(n_clusters=3, max_iter=10000, random_state=0)
km_cluster.fit(feature_vect)
cluster_label = km_cluster.labels_

# 소속 클러스터를 cluster_label 칼럼으로 할당하고 cluster_label 값으로 정렬
document_df['cluster_label'] = cluster_label
document_df.sort_values(by='cluster_label')

filename  \
25          rooms_swissotel_chicago   
26    gas_mileage_toyota_camry_2007   
27       interior_toyota_camry_2007   
31        comfort_toyota_camry_2007   
32        mileage_honda_accord_2008   
13   location_bestwestern_hotel_sfo   
12      rooms_bestwestern_hotel_sfo   
11      location_holiday_inn_london   
49         price_holiday_inn_london   
10   transmission_toyota_camry_2007   
8         quality_toyota_camry_2007   
39   bathroom_bestwestern_hotel_sfo   
40          seats_honda_accord_2008   
41    parking_bestwestern_hotel_sfo   
47    performance_honda_accord_2008   
48       interior_honda_accord_2008   
38        comfort_honda_accord_2008   
23          room_holiday_inn_london   
36              sound_ipod_nano_8gb   
44              video_ipod_nano_8gb   
45       performance_netbook_1005ha   
30       battery-life_ipod_nano_8gb   
50          keyboard_netbook_1005ha   
3             screen_netbook_1005ha   
4              screen_ipod_nano_8gb   
17      battery-life_netbook_1005ha   
16       battery-life_amazon_kindle   
5          size_asus_netbook_1005ha   
21     updates_garmin_nuvi_255W_gps   
1      display_garmin_nuvi_255W_gps   
2    satellite_garmin_nuvi_255W_gps   
46  service_swissotel_hotel_chicago   
43       free_bestwestern_hotel_sfo   
42          food_holiday_inn_london   
6        voice_garmin_nuvi_255W_gps   
7       screen_garmin_nuvi_255W_gps   
9           staff_swissotel_chicago   
14    eyesight-issues_amazon_kindle   
22       service_holiday_inn_london   
35                   speed_windows7   
34           food_swissotel_chicago   
33         navigation_amazon_kindle   
15              fonts_amazon_kindle   
29              price_amazon_kindle   
28                features_windows7   
18    service_bestwestern_hotel_sfo   
19            buttons_amazon_kindle   
20    accuracy_garmin_nuvi_255W_gps   
24       speed_garmin_nuvi_255W_gps   
37      staff_bestwestern_hotel_sfo   
0   directions_garmin_nuvi_255W_gps   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
25                                                                                                                                                                                                                    The Swissotel is one of our favorite hotels in Chicago and the corner rooms have the most fantastic views in the city .\n0                                                                                                                                   The rooms look like they were just remodled and upgraded, there was an HD TV and a nice iHome docking station to put my iPod so I could set the alarm to wake up with my music instead of the radio .\n1                                 ...   
26                                                                                                                                                                                                                                                                                                                                                                              Ride seems comfortable and gas mileage fairly good averaging 26 city and 30 open road .\n0                                                                                                                              

### **군집별 핵심 단어 추출하기**
- KMeans 객체는 각 군집을 구성하는 단어 피처가 군집의 중심(Centroid)을 기준으로 얼마나 가깝게 위치해있는지 `clusters_centers_` 속성으로 제공
  - 배열 값으로 제공
  - 행: 개별 군집, 열: 개별 피처
  - 개별 군집 내의 상대 위치를 숫자 값으로 표현한 일종의 좌표 값

In [24]:
cluster_centers = km_cluster.cluster_centers_
print('cluster_centers shape:', cluster_centers.shape)
print(cluster_centers)

cluster_centers shape: (3, 4611)
[[0.         0.00139861 0.00105506 ... 0.         0.00102516 0.00014966]
 [0.00973837 0.         0.         ... 0.00397274 0.         0.        ]
 [0.00669334 0.         0.00038917 ... 0.00594976 0.00047351 0.00088866]]


- shape: (3, 4611)
  - 군집: 3개, word 피처: 4611개
  - 배열 값: 0~1까지의 값, 1에 가까울수록 중심과 가까움
- `argsort()[:, ::-]` -> 큰 순으로 정렬된 위치 인덱스 값 반환

In [25]:
# 군집별 top n 핵심단어, 그 단어의 중심 위치 상대값, 대상 파일명들을 반환함.
def get_cluster_details(cluster_model, cluster_data, feature_names, clusters_num, top_n_features=10):
    cluster_details = {}

    # cluster_centers array 의 값이 큰 순으로 정렬된 index 값을 반환
    # 군집 중심점(centroid)별 할당된 word 피처들의 거리값이 큰 순으로 값을 구하기 위함.
    centroid_feature_ordered_ind = cluster_model.cluster_centers_.argsort()[:,::-1]

    #개별 군집별로 iteration하면서 핵심단어, 그 단어의 중심 위치 상대값, 대상 파일명 입력
    for cluster_num in range(clusters_num):
        # 개별 군집별 정보를 담을 데이터 초기화.
        cluster_details[cluster_num] = {}
        cluster_details[cluster_num]['cluster'] = cluster_num

        # cluster_centers_.argsort()[:,::-1] 로 구한 index 를 이용하여 top n 피처 단어를 구함.
        top_feature_indexes = centroid_feature_ordered_ind[cluster_num, :top_n_features]
        top_features = [ feature_names[ind] for ind in top_feature_indexes ]

        # top_feature_indexes를 이용해 해당 피처 단어의 중심 위치 상댓값 구함
        top_feature_values = cluster_model.cluster_centers_[cluster_num, top_feature_indexes].tolist()

        # cluster_details 딕셔너리 객체에 개별 군집별 핵심 단어와 중심위치 상대값, 그리고 해당 파일명 입력
        cluster_details[cluster_num]['top_features'] = top_features
        cluster_details[cluster_num]['top_features_value'] = top_feature_values
        filenames = cluster_data[cluster_data['cluster_label'] == cluster_num]['filename']
        filenames = filenames.values.tolist()
        cluster_details[cluster_num]['filenames'] = filenames

    return cluster_details

In [26]:
def print_cluster_details(cluster_details):
    for cluster_num, cluster_detail in cluster_details.items():
        print('####### Cluster {0}'.format(cluster_num))
        print('Top features:', cluster_detail['top_features'])
        print('Reviews 파일명 :',cluster_detail['filenames'][:7])
        print('==================================================')

In [28]:
feature_names = tfidf_vect.get_feature_names_out()
cluster_details = get_cluster_details(cluster_model=km_cluster, cluster_data=document_df,
                                      feature_names=feature_names, clusters_num=3, top_n_features=10)
print_cluster_details(cluster_details)

####### Cluster 0
Top features: ['room', 'interior', 'seat', 'mileage', 'hotel', 'location', 'comfortable', 'gas', 'car', 'gas mileage']
Reviews 파일명 : ['quality_toyota_camry_2007', 'transmission_toyota_camry_2007', 'location_holiday_inn_london', 'rooms_bestwestern_hotel_sfo', 'location_bestwestern_hotel_sfo', 'room_holiday_inn_london', 'rooms_swissotel_chicago']
####### Cluster 1
Top features: ['battery', 'screen', 'battery life', 'life', 'keyboard', 'video', 'size', 'performance', 'sound', 'ipod']
Reviews 파일명 : ['screen_netbook_1005ha', 'screen_ipod_nano_8gb', 'size_asus_netbook_1005ha', 'battery-life_amazon_kindle', 'battery-life_netbook_1005ha', 'battery-life_ipod_nano_8gb', 'sound_ipod_nano_8gb']
####### Cluster 2
Top features: ['service', 'staff', 'food', 'hotel', 'screen', 'direction', 'voice', 'map', 'room', 'kindle']
Reviews 파일명 : ['directions_garmin_nuvi_255W_gps', 'display_garmin_nuvi_255W_gps', 'satellite_garmin_nuvi_255W_gps', 'voice_garmin_nuvi_255W_gps', 'screen_garmin_nu

- Cluster 0: 호텔 리뷰 군집과 자동차 리뷰 군집 일부가 섞임
  - 방과 서비스 등 위주
- Cluster 1: 포터블 전자제품 리뷰 군집
  - 화면, 배터리 수명 등
- Cluster 2: 자동차 리뷰 군집과 호텔 리뷰 군집 일부가 섞임